test/unittests/parser_test.cpp

```diff
@@ -1241,10 +1243,23 @@ TEST(parser, data_section_memidx_nonzero)
 {
     const auto section_contents = make_vec({"0141010b0100"_bytes});
     const auto bin = bytes{wasm_prefix} + make_section(11, section_contents);
-
     EXPECT_THROW_MESSAGE(parse(bin), parser_error, "unexpected memidx value 1");
 }
 
+TEST(parser, DISABLED_empty_data_section_without_memory)
+{
+    const auto bin = bytes{wasm_prefix} + make_section(11, make_vec({}));
```

Comment:
Since empty data sections end up being `datasec.empty()` in the parser, this case is not detected.

@gumb0 did you say the spec states an empty data section can or cannot be present if no memory section is present?

Reply:
It cannot. Any data section is invalid if no memory is present, according to validation spec.

![image](https://user-images.githubusercontent.com/1863135/82999997-16bfb100-a00a-11ea-9fba-27f6a2fa71cd.png)

Reply:
Any suggestion how to clearly support that in our code? Since section kind's are strictly ordered (and we enforce that already), the best option seems to be moving the check into the switch statement right before `parse_vec<Data>`. What do you think?

Reply:
It requires info about imports, so to move data and element section checks into the switch, you would need to move "split imports" loop inside `case SectionId::import`, too.

I'm fine in this PR to leave like it is now.

Reply:
The current check covers empty data section, too, doesn't it? So this test can be just enabled?

Reply:
> The current check covers empty data section, too, doesn't it? So this test can be just enabled?

It does not, if it is enabled it fails. You can try yourself and perhaps come up with a clever solution.

Reply:
Ah, I misunderstood what is '"empty data section", I thought it's "data section with empty data vector" (that would be invalid), but you mean "section consisting of empty vector". I can't find any restriction for a data section as a whole, I guess any section with empty contents is allowed.

Also, in binary spec it says:
![image](https://user-images.githubusercontent.com/1863135/83008482-825b4b80-a015-11ea-92d8-2364e73513dc.png)

Reply:
Good find, will take your change then.